### PR TITLE
mizani, plotnine, and pystan: added more versions and variable dependencies

### PR DESCRIPTION
**Pull request by luke-dt**

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-mizani/package.py [hunk 1]
```diff
@@ -2,23 +2,38 @@
...
+    version('0.7.3', sha256='f521300bd29ca918fcd629bc8ab50fa04e41bdbe00f6bcf74055d3c6273770a4')
+    version('0.6.0', sha256='2cdba487ee54faf3e5bfe0903155a13ff13d27a2dae709f9432194915b4fb1cd')
 
-    depends_on('python@3.6:', type=('build', 'run'))
-    depends_on("py-matplotlib@3.1.1:", type=("build", "run"))
-    depends_on("py-numpy", type=("build", "run"))
-    depends_on("py-palettable", type=("build", "run"))
-    depends_on("py-pandas@1.1.0:", type=("build", "run"))
+    # common requirements
     depends_on("py-setuptools", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-mizani/package.py]: 42+ required now

File: var/spack/repos/builtin/packages/py-mizani/package.py [hunk 2]
```diff
@@ -2,23 +2,38 @@
...
+
+    depends_on("py-matplotlib@3.5.0:", type=("build", "run"), when='@0.7.4:')
+    depends_on("py-matplotlib@3.1.1:", type=("build", "run"), when='@0.6.0:')
+
+    depends_on("py-numpy@1.19.0:", type=("build", "run"), when='@0.7.4:')
+    depends_on("py-numpy",         type=("build", "run"), when='@0.6.0:')
+
+    depends_on("py-scipy@1.5.0:", type=("build", "run"), when='@0.7.4:')
+
+    depends_on("py-pandas@1.1.0:",  type=("build", "run"), when='@0.7.2:')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-mizani/package.py:40]: 1.3.5+ required now

File: var/spack/repos/builtin/packages/py-plotnine/package.py [hunk 3]
```diff
@@ -10,20 +12,44 @@ class PyPlotnine(PythonPackage):
...
+    pypi     = "plotnine/plotnine-0.8.0.tar.gz"
+
+    version('0.9.0', sha256='0e89a93015f3c71d6844ac7aa9fb0da09b908f5f7dfa7dd5d68a5ca32b2ebcea')
+    version('0.8.0', sha256='39de59edcc28106761b65238647d0b1f6212ea7f3a78f8be0b846616db969276')
+    version('0.7.1', sha256='02f2b0435dae2e917198c5367fd97b010445d64d9888c6b7e755d3cdfe7ad057')
+    version('0.7.0', sha256='8ee67cbf010ccea32670760e930b7b02177030a89ccdf85e35d156a96ce36cd3')
+    version('0.6.0', sha256='aae2c8164abb209ef4f28cab01132d23f6879fcf8d492657487359e1241459e5')
+
+    # common requirements
+    depends_on("py-setuptools",          type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-plotnine/package.py]: Newer versions needed:
```toml
[build-system]
requires = [
    "setuptools>=59",
    "setuptools_scm[toml]>=6.4",
]
```

File: var/spack/repos/builtin/packages/py-plotnine/package.py [hunk 4]
```diff
@@ -10,20 +12,44 @@ class PyPlotnine(PythonPackage):
...
+    depends_on("py-matplotlib@3.5.0:",   type=("build", "run"), when='@0.9.0:')
+    depends_on("py-matplotlib@3.1.1:",   type=("build", "run"), when='@0.6.0:')
+
+    depends_on("py-mizani@0.7.3:",       type=("build", "run"), when='@0.8.0:')
+    depends_on("py-mizani@0.6.0:",       type=("build", "run"), when='@0.6.0:')
+
+    depends_on("py-numpy@1.19.0:",       type=("build", "run"), when='@0.8.0:')
+    depends_on("py-numpy@1.16.0:",       type=("build", "run"), when='@0.6.0:')
+
+    depends_on("py-pandas@1.1.0:",       type=("build", "run"), when='@0.7.1:')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-plotnine/package.py]: 1.3.5+ required now

File: var/spack/repos/builtin/packages/py-pystan/package.py [hunk 5]
```diff
@@ -14,13 +15,23 @@ class PyPystan(PythonPackage):
...
-    version('3.4.0', sha256='325e2fb0ab804555c05a603e0c9152ab11fcc3af01f3e9a9ff9fe9954b93184f')
+    version('3.5.0',    sha256='078571d071a5b7c0af59206d4994a0979f4ac4b61f4a720b640c44fe35514929')
+    version('3.4.0',    sha256='325e2fb0ab804555c05a603e0c9152ab11fcc3af01f3e9a9ff9fe9954b93184f')
+    version('2.19.1.1', sha256='fa8bad8dbc0da22bbe6f36af56c9abbfcf10f92df8ce627d59a36bd8d25eb038')
+    version('2.19.0.0', sha256='b85301b960d5991918b40bd64a4e9321813657a9fc028e0f39edce7220a309eb')
 
-    depends_on('python@3.8:3', type=('build', 'run'))
-    depends_on('py-setuptools', type=('build', 'run'))
+    # common requirements
+    depends_on('py-setuptools',         type='build')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pystan/package.py]: This should probably be type=build/run

File: var/spack/repos/builtin/packages/py-pystan/package.py [hunk 6]
```diff
@@ -14,13 +15,23 @@ class PyPystan(PythonPackage):
...
-    depends_on('py-numpy@1.19:1', type=('build', 'run'))
-    depends_on('py-clikit@0.6', type=('build', 'run'))
+
+    # variable requirements
+    depends_on('python@3.8:3.9',    type=('build', 'run'), when='@3.4.0:')
+    depends_on('py-aiohttp@3.6:',   type=('build', 'run'), when='@3.4.0:')
+    depends_on('py-httpstan@4.8:',  type=('build', 'run'), when='@3.5.0:')
+    depends_on('py-httpstan@4.7',   type=('build', 'run'), when='@3.4.0:')
+    depends_on('py-pysimdjson@3.2', type=('build', 'run'), when='@3.4.0:')
+    depends_on('py-numpy@1.19:',    type=('build', 'run'), when='@3.4.0:')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pystan/package.py]:
```suggestion
    depends_on('python@3.8:3',    type=('build', 'run'), when='@3.4.0:')
    depends_on('py-aiohttp@3.6:3',   type=('build', 'run'), when='@3.4.0:')
    depends_on('py-httpstan@4.8',  type=('build', 'run'), when='@3.5.0:')
    depends_on('py-httpstan@4.7',   type=('build', 'run'), when='@3.4')
    depends_on('py-pysimdjson@3.2:3', type=('build', 'run'), when='@3.4.0:')
    depends_on('py-numpy@1.19:1',    type=('build', 'run'), when='@3.4.0:')
```
See https://python-poetry.org/docs/dependency-specification/

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-pystan/package.py (hunk 6 above)
Comment (luke-dt) [reply var/spack/repos/builtin/packages/py-pystan/package.py]: thanks for the link! I wasn't aware of the caret/tidle specs

File: var/spack/repos/builtin/packages/py-plotnine/package.py [hunk 7]
```diff
@@ -10,20 +12,44 @@ class PyPlotnine(PythonPackage):
...
+    version('0.8.0', sha256='39de59edcc28106761b65238647d0b1f6212ea7f3a78f8be0b846616db969276')
+    version('0.7.1', sha256='02f2b0435dae2e917198c5367fd97b010445d64d9888c6b7e755d3cdfe7ad057')
+    version('0.7.0', sha256='8ee67cbf010ccea32670760e930b7b02177030a89ccdf85e35d156a96ce36cd3')
+    version('0.6.0', sha256='aae2c8164abb209ef4f28cab01132d23f6879fcf8d492657487359e1241459e5')
+
+    depends_on('python@3.8:',            type=('build', 'run'), when='@0.9.0:')
+    depends_on('python@3.6:',            type=('build', 'run'), when='@0.6.0:')
+
+    depends_on("py-setuptools@59:",      type=("build", "run"), when='@0.9.0:')
+    depends_on("py-setuptools",          type=("build", "run"), when='@0.6.0:')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-plotnine/package.py]: I don't think this is needed at run-time. You're also missing a build dep on `py-setuptools-scm@6.4:+toml`

File: var/spack/repos/builtin/packages/py-pystan/package.py (hunk 5 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-pystan/package.py]: Should be type=build/run

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-pystan/package.py (hunk 5 above)
Comment (luke-dt) [reply var/spack/repos/builtin/packages/py-pystan/package.py]: oh, I misread this comment as for py-plotnine.

**Review by adamjstewart — DISMISSED** [review]

**Comment by adamjstewart** [comment]
Looks like you'll need to add new versions of statsmodels and httpstan:
```
1. py-plotnine: dependency on py-statsmodels@0.13.2: cannot be satisfied by known versions of py-statsmodels
    happening in /home/runner/work/spack/spack/var/spack/repos/builtin/packages/py-plotnine/package.py
    known versions of py-statsmodels are 0.13.1, 0.13.0, 0.12.2, 0.12.1, 0.10.2, 0.10.1, 0.8.0
2. py-pystan: dependency on py-httpstan@4.8 cannot be satisfied by known versions of py-httpstan
    happening in /home/runner/work/spack/spack/var/spack/repos/builtin/packages/py-pystan/package.py
    known versions of py-httpstan are 4.7.2, 4.6.1
```

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**